### ISSUE TITLE
Cannot download XLSX if there's more than 1 million results

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#13585]
**Describe the bug**
It's not possible to download XLSX export from a result with 1048576 or more rows, since the limitation in XLSX standard is 1048576 rows, but Metabase downloads 1048576+1 (header row). Limiting the results to 1048575 will work.
Alternative workaround is to download as CSV.

**To Reproduce**
1. Custom question > Sample Dataset > People
2. Join People again on Source=Source
![image](https://user-images.githubusercontent.com/1447303/97042846-ef95ed00-1571-11eb-8733-6f3f3b062c9a.png)
3. Export XLSX - the downloaded file is corrupted with a stacktrace - error:
`Invalid row number (1048576) outside allowable range (0..1048575)`

<details><summary>Full stacktrace</summary>

```
[ccfc9f9e-21e3-42a0-95a0-6f29bf142ba9] 2020-10-23T20:05:39+02:00 ERROR metabase.query-processor.middleware.catch-exceptions Error processing query: null
{:database_id 1,
 :started_at #t "2020-10-23T18:04:34.310538Z[GMT]",
 :via
 [{:status :failed,
   :class clojure.lang.ExceptionInfo,
   :error "Error reducing result rows",
   :stacktrace
   ["--> query_processor.context.default$default_reducef$fn__38214.invoke(default.clj:61)"
    "query_processor.context.default$default_reducef.invokeStatic(default.clj:58)"
    "query_processor.context.default$default_reducef.invoke(default.clj:49)"
    "query_processor.context$reducef.invokeStatic(context.clj:69)"
    "query_processor.context$reducef.invoke(context.clj:62)"
    "query_processor.context.default$default_runf$respond_STAR___38218.invoke(default.clj:70)"
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:392)"
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)"
    "driver.sql_jdbc$fn__77417.invokeStatic(sql_jdbc.clj:49)"
    "driver.sql_jdbc$fn__77417.invoke(sql_jdbc.clj:47)"
    "driver.h2$fn__76231.invokeStatic(h2.clj:87)"
    "driver.h2$fn__76231.invoke(h2.clj:84)"
    "query_processor.context$executef.invokeStatic(context.clj:59)"
    "query_processor.context$executef.invoke(context.clj:48)"
    "query_processor.context.default$default_runf.invokeStatic(default.clj:69)"
    "query_processor.context.default$default_runf.invoke(default.clj:67)"
    "query_processor.context$runf.invokeStatic(context.clj:45)"
    "query_processor.context$runf.invoke(context.clj:39)"
    "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
    "query_processor.reducible$pivot.invoke(reducible.clj:31)"
    "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__45926.invoke(mbql_to_native.clj:26)"
    "query_processor.middleware.check_features$check_features$fn__45201.invoke(check_features.clj:42)"
    "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__46091.invoke(optimize_datetime_filters.clj:133)"
    "query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__44006.invoke(auto_parse_filter_values.clj:44)"
    "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__47657.invoke(wrap_value_literals.clj:149)"
    "query_processor.middleware.annotate$add_column_info$fn__43768.invoke(annotate.clj:574)"
    "query_processor.middleware.permissions$check_query_permissions$fn__45076.invoke(permissions.clj:64)"
    "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__46609.invoke(pre_alias_aggregations.clj:40)"
    "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__45274.invoke(cumulative_aggregations.clj:61)"
    "query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__46846.invoke(resolve_joined_fields.clj:36)"
    "query_processor.middleware.resolve_joins$resolve_joins$fn__47165.invoke(resolve_joins.clj:183)"
    "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__39493.invoke(add_implicit_joins.clj:254)"
    "query_processor.middleware.large_int_id$convert_id_to_string$fn__45887.invoke(large_int_id.clj:44)"
    "query_processor.middleware.limit$limit$fn__45912.invoke(limit.clj:38)"
    "query_processor.middleware.format_rows$format_rows$fn__45867.invoke(format_rows.clj:84)"
    "query_processor.middleware.desugar$desugar$fn__45340.invoke(desugar.clj:22)"
    "query_processor.middleware.binning$update_binning_strategy$fn__44365.invoke(binning.clj:229)"
    "query_processor.middleware.resolve_fields$resolve_fields$fn__44882.invoke(resolve_fields.clj:24)"
    "query_processor.middleware.add_dimension_projections$add_remapping$fn__39037.invoke(add_dimension_projections.clj:318)"
    "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__39244.invoke(add_implicit_clauses.clj:141)"
    "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__39642.invoke(add_source_metadata.clj:105)"
    "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__46806.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)"
    "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__43953.invoke(auto_bucket_datetimes.clj:125)"
    "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__44929.invoke(resolve_source_table.clj:46)"
    "query_processor.middleware.parameters$substitute_parameters$fn__46591.invoke(parameters.clj:114)"
    "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__44981.invoke(resolve_referenced.clj:80)"
    "query_processor.middleware.expand_macros$expand_macros$fn__45596.invoke(expand_macros.clj:158)"
    "query_processor.middleware.add_timezone_info$add_timezone_info$fn__39673.invoke(add_timezone_info.clj:15)"
    "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__47527.invoke(splice_params_in_response.clj:32)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46817$fn__46821.invoke(resolve_database_and_driver.clj:33)"
    "driver$do_with_driver.invokeStatic(driver.clj:61)"
    "driver$do_with_driver.invoke(driver.clj:57)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46817.invoke(resolve_database_and_driver.clj:27)"
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__45814.invoke(fetch_source_query.clj:267)"
    "query_processor.middleware.store$initialize_store$fn__47536$fn__47537.invoke(store.clj:11)"
    "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
    "query_processor.store$do_with_store.invoke(store.clj:40)"
    "query_processor.middleware.store$initialize_store$fn__47536.invoke(store.clj:10)"
    "query_processor.middleware.cache$maybe_return_cached_results$fn__44858.invoke(cache.clj:214)"
    "query_processor.middleware.validate$validate_query$fn__47545.invoke(validate.clj:10)"
    "query_processor.middleware.normalize_query$normalize$fn__45939.invoke(normalize_query.clj:22)"
    "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__39511.invoke(add_rows_truncated.clj:36)"
    "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__47512.invoke(results_metadata.clj:143)"
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__45217.invoke(constraints.clj:42)"
    "query_processor.middleware.process_userland_query$process_userland_query$fn__46680.invoke(process_userland_query.clj:136)"
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__45160.invoke(catch_exceptions.clj:174)"
    "query_processor.reducible$async_qp$qp_STAR___38300$thunk__38301.invoke(reducible.clj:101)"
    "query_processor.reducible$async_qp$qp_STAR___38300$fn__38303.invoke(reducible.clj:106)"],
   :error_type :qp,
   :ex-data {:type :qp}}],
 :error_type :qp,
 :json_query
 {:database 1,
  :query {:source-table 4, :joins [{:fields "none", :source-table 4, :condition ["=" ["field-id" 6] ["joined-field" "People" ["field-id" 6]]], :alias "People"}], :filter ["<" ["field-id" 30] 2098]},
  :type "query",
  :middleware {:skip-results-metadata? true, :format-rows? false},
  :async? true},
 :native
 {:query
  "SELECT \"PUBLIC\".\"PEOPLE\".\"ID\" AS \"ID\", \"PUBLIC\".\"PEOPLE\".\"ADDRESS\" AS \"ADDRESS\", \"PUBLIC\".\"PEOPLE\".\"EMAIL\" AS \"EMAIL\", \"PUBLIC\".\"PEOPLE\".\"NAME\" AS \"NAME\", \"PUBLIC\".\"PEOPLE\".\"CITY\" AS \"CITY\", \"PUBLIC\".\"PEOPLE\".\"LONGITUDE\" AS \"LONGITUDE\", \"PUBLIC\".\"PEOPLE\".\"STATE\" AS \"STATE\", \"PUBLIC\".\"PEOPLE\".\"SOURCE\" AS \"SOURCE\", \"PUBLIC\".\"PEOPLE\".\"BIRTH_DATE\" AS \"BIRTH_DATE\", \"PUBLIC\".\"PEOPLE\".\"ZIP\" AS \"ZIP\", \"PUBLIC\".\"PEOPLE\".\"LATITUDE\" AS \"LATITUDE\", \"PUBLIC\".\"PEOPLE\".\"CREATED_AT\" AS \"CREATED_AT\" FROM \"PUBLIC\".\"PEOPLE\" LEFT JOIN \"PUBLIC\".\"PEOPLE\" \"People\" ON \"PUBLIC\".\"PEOPLE\".\"SOURCE\" = \"People\".\"SOURCE\" WHERE \"PUBLIC\".\"PEOPLE\".\"ID\" < 2098 LIMIT 1048576",
  :params nil},
 :status :failed,
 :class java.lang.IllegalArgumentException,
 :stacktrace
 ["org.apache.poi.xssf.streaming.SXSSFSheet.createRow(SXSSFSheet.java:123)"
  "org.apache.poi.xssf.streaming.SXSSFSheet.createRow(SXSSFSheet.java:65)"
  "dk.ative.docjure.spreadsheet$add_row_BANG_.invokeStatic(spreadsheet.clj:265)"
  "dk.ative.docjure.spreadsheet$add_row_BANG_.invoke(spreadsheet.clj:260)"
  "--> query_processor.streaming.xlsx$fn$reify__35650.write_row_BANG_(xlsx.clj:107)"
  "query_processor.streaming$streaming_rff$fn__35657$fn__35658.invoke(streaming.clj:35)"
  "query_processor.middleware.process_userland_query$add_and_save_execution_info_xform_BANG_$execution_info_rf_STAR___46671.invoke(process_userland_query.clj:95)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated_xform$fn__39506.invoke(add_rows_truncated.clj:28)"
  "query_processor.middleware.limit$limit_xform$fn__45909.invoke(limit.clj:25)"
  "query_processor.reducible$reducible_rows$reify__38316.reduce(reducible.clj:157)"
  "query_processor.context.default$default_reducef$fn__38214.invoke(default.clj:59)"
  "query_processor.context.default$default_reducef.invokeStatic(default.clj:58)"
  "query_processor.context.default$default_reducef.invoke(default.clj:49)"
  "query_processor.context$reducef.invokeStatic(context.clj:69)"
  "query_processor.context$reducef.invoke(context.clj:62)"
  "query_processor.context.default$default_runf$respond_STAR___38218.invoke(default.clj:70)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:392)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)"
  "driver.sql_jdbc$fn__77417.invokeStatic(sql_jdbc.clj:49)"
  "driver.sql_jdbc$fn__77417.invoke(sql_jdbc.clj:47)"
  "driver.h2$fn__76231.invokeStatic(h2.clj:87)"
  "driver.h2$fn__76231.invoke(h2.clj:84)"
  "query_processor.context$executef.invokeStatic(context.clj:59)"
  "query_processor.context$executef.invoke(context.clj:48)"
  "query_processor.context.default$default_runf.invokeStatic(default.clj:69)"
  "query_processor.context.default$default_runf.invoke(default.clj:67)"
  "query_processor.context$runf.invokeStatic(context.clj:45)"
  "query_processor.context$runf.invoke(context.clj:39)"
  "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
  "query_processor.reducible$pivot.invoke(reducible.clj:31)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__45926.invoke(mbql_to_native.clj:26)"
  "query_processor.middleware.check_features$check_features$fn__45201.invoke(check_features.clj:42)"
  "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__46091.invoke(optimize_datetime_filters.clj:133)"
  "query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__44006.invoke(auto_parse_filter_values.clj:44)"
  "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__47657.invoke(wrap_value_literals.clj:149)"
  "query_processor.middleware.annotate$add_column_info$fn__43768.invoke(annotate.clj:574)"
  "query_processor.middleware.permissions$check_query_permissions$fn__45076.invoke(permissions.clj:64)"
  "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__46609.invoke(pre_alias_aggregations.clj:40)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__45274.invoke(cumulative_aggregations.clj:61)"
  "query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__46846.invoke(resolve_joined_fields.clj:36)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__47165.invoke(resolve_joins.clj:183)"
  "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__39493.invoke(add_implicit_joins.clj:254)"
  "query_processor.middleware.large_int_id$convert_id_to_string$fn__45887.invoke(large_int_id.clj:44)"
  "query_processor.middleware.limit$limit$fn__45912.invoke(limit.clj:38)"
  "query_processor.middleware.format_rows$format_rows$fn__45867.invoke(format_rows.clj:84)"
  "query_processor.middleware.desugar$desugar$fn__45340.invoke(desugar.clj:22)"
  "query_processor.middleware.binning$update_binning_strategy$fn__44365.invoke(binning.clj:229)"
  "query_processor.middleware.resolve_fields$resolve_fields$fn__44882.invoke(resolve_fields.clj:24)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__39037.invoke(add_dimension_projections.clj:318)"
  "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__39244.invoke(add_implicit_clauses.clj:141)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__39642.invoke(add_source_metadata.clj:105)"
  "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__46806.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)"
  "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__43953.invoke(auto_bucket_datetimes.clj:125)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__44929.invoke(resolve_source_table.clj:46)"
  "query_processor.middleware.parameters$substitute_parameters$fn__46591.invoke(parameters.clj:114)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__44981.invoke(resolve_referenced.clj:80)"
  "query_processor.middleware.expand_macros$expand_macros$fn__45596.invoke(expand_macros.clj:158)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__39673.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__47527.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46817$fn__46821.invoke(resolve_database_and_driver.clj:33)"
  "driver$do_with_driver.invokeStatic(driver.clj:61)"
  "driver$do_with_driver.invoke(driver.clj:57)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46817.invoke(resolve_database_and_driver.clj:27)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__45814.invoke(fetch_source_query.clj:267)"
  "query_processor.middleware.store$initialize_store$fn__47536$fn__47537.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
  "query_processor.store$do_with_store.invoke(store.clj:40)"
  "query_processor.middleware.store$initialize_store$fn__47536.invoke(store.clj:10)"
  "query_processor.middleware.cache$maybe_return_cached_results$fn__44858.invoke(cache.clj:214)"
  "query_processor.middleware.validate$validate_query$fn__47545.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__45939.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__39511.invoke(add_rows_truncated.clj:36)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__47512.invoke(results_metadata.clj:143)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__45217.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__46680.invoke(process_userland_query.clj:136)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__45160.invoke(catch_exceptions.clj:174)"
  "query_processor.reducible$async_qp$qp_STAR___38300$thunk__38301.invoke(reducible.clj:101)"
  "query_processor.reducible$async_qp$qp_STAR___38300$fn__38303.invoke(reducible.clj:106)"],
 :context :xlsx-download,
 :error "Invalid row number (1048576) outside allowable range (0..1048575)",
 :row_count 0,
 :running_time 0,
 :preprocessed
 {:database 1,
  :query
  {:source-table 4,
   :joins [{:strategy :left-join, :source-table 4, :condition [:= [:field-id 6] [:joined-field "People" [:field-id 6]]], :alias "People"}],
   :filter [:< [:field-id 30] [:value 2098 {:base_type :type/BigInteger, :special_type :type/PK, :database_type "BIGINT", :name "ID"}]],
   :fields
   [[:field-id 30]
    [:field-id 32]
    [:field-id 34]
    [:field-id 11]
    [:field-id 13]
    [:field-id 31]
    [:field-id 33]
    [:field-id 6]
    [:datetime-field [:field-id 17] :default]
    [:field-id 26]
    [:field-id 21]
    [:datetime-field [:field-id 3] :default]],
   :limit 1048576},
  :type :query,
  :middleware {:skip-results-metadata? true, :format-rows? false},
  :async? true,
  :info
  {:executed-by 44,
   :context :xlsx-download,
   :query-hash [-119, -55, -122, -49, -128, -19, 89, -59, -74, 96, 0, 102, -105, -79, 63, 42, 104, 44, 86, -36, -74, 8, -47, -99, 106, -12, -39, -47, 44, -74, -95, 46]}},
 :data {:rows [], :cols []}}

[ccfc9f9e-21e3-42a0-95a0-6f29bf142ba9] 2020-10-23T20:05:39+02:00 DEBUG metabase.middleware.log POST /api/dataset/xlsx 202 [ASYNC: completed] 1.1 mins (12 DB calls) App DB connections: 0/15 Jetty threads: 2/50 (12 idle, 0 queued) (120 total active threads) Queries in flight: 1 (0 queued)
```

</details>

**Information about your Metabase Installation:**
Master 0.37.0 (it might have been a problem before, but it's killing my computer to run the test, so haven't tried earlier)

**Additional context**
Related to #4564, since even if we implemented a higher number of rows, then XLSX still needs this limitation.
Also related to #10566, since a download this size will take a lot of time, so some indication would be helpful.

**Comment by paoliniluis** [comment]
Just run into this one: I would suggest that if the query returns more than 1048575 rows (I think we know that beforehand) we should disable the excel export button altogether to save resources and prevent the error.
I don't think that excel will raise the limit of their own format, and most power users of Excel recommend going straight to CSV when handling lots of data and reducing it to a smaller subset of data with Excel's add-ins.

**Comment by camsaul** [comment]
@paoliniluis we can just limit the query to the number of rows we want

**Comment by camsaul** [comment]
Fixed by #15414
